File: android/src/main/java/expo/modules/pdfhelpers/ReactNativePdfHelpersModule.kt
package expo.modules.pdfhelpers

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.content.ContentResolver
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.Build
import android.os.ParcelFileDescriptor
import androidx.annotation.RequiresApi
import com.facebook.react.bridge.*
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

class ReactNativePdfHelpersModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ReactNativePdfHelpers')` in JavaScript.
    Name("ReactNativePdfHelpers")

    AsyncFunction("getPageCount") { filePath: String, promise: Promise ->
      var parcelFileDescriptor: ParcelFileDescriptor? = null
      var pdfRenderer: PdfRenderer? = null
      try {
        parcelFileDescriptor = getParcelFileDescriptor(filePath)
        if (parcelFileDescriptor == null) {
          promise.reject("FILE_NOT_FOUND", "File $filePath not found")
          return
        }

        pdfRenderer = PdfRenderer(parcelFileDescriptor)

        promise.resolve(pdfRenderer.pageCount)
      } catch (ex: IOException) {
        promise.reject("INTERNAL_ERROR", ex)
      } finally {
        pdfRenderer?.close()
        parcelFileDescriptor?.close()
      }
    }

    AsyncFunction("generateThumbnail") { filePath: String, page: Int, quality: Int, promise: Promise ->
      var parcelFileDescriptor: ParcelFileDescriptor? = null
      var pdfRenderer: PdfRenderer? = null
      try {
        parcelFileDescriptor = getParcelFileDescriptor(filePath)
        if (parcelFileDescriptor == null) {
          promise.reject("FILE_NOT_FOUND", "File $filePath not found")
          return
        }

        pdfRenderer = PdfRenderer(parcelFileDescriptor)
        if (page < 0 || page >= pdfRenderer.pageCount) {
          promise.reject("INVALID_PAGE", "Page number $page is invalid, file has ${pdfRenderer.pageCount} pages")
          return
        }

        val result = renderPage(pdfRenderer, page, filePath, quality)
        promise.resolve(result)
      } catch (ex: IOException) {
        promise.reject("INTERNAL_ERROR", ex)
      } finally {
        pdfRenderer?.close()
        parcelFileDescriptor?.close()
      }
    }
  }

  private fun getParcelFileDescriptor(filePath: String): ParcelFileDescriptor? {
    val uri = Uri.parse(filePath)
    if (ContentResolver.SCHEME_CONTENT == uri.scheme || ContentResolver.SCHEME_FILE == uri.scheme) {
      return this.reactApplicationContext.contentResolver.openFileDescriptor(uri, "r")
    } else if (filePath.startsWith("/")) {
      val file = File(filePath)
      return ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
    }
    return null
  }

  private fun renderPage(pdfRenderer: PdfRenderer, page: Int, filePath: String, quality: Int): WritableNativeMap {
    val currentPage = pdfRenderer.openPage(page)
    val width = currentPage.width
    val height = currentPage.height
    val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
    currentPage.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
    currentPage.close()

    // Some bitmaps have transparent background which results in a black thumbnail. Add a white background.
    val bitmapWhiteBG = Bitmap.createBitmap(bitmap.width, bitmap.height, bitmap.config)
    bitmapWhiteBG.eraseColor(Color.WHITE)
    val canvas = Canvas(bitmapWhiteBG)
    canvas.drawBitmap(bitmap, 0f, 0f, null)
    bitmap.recycle()

    val outputFile = File.createTempFile(getOutputFilePrefix(filePath, page), ".jpg", reactApplicationContext.cacheDir)
    if (outputFile.exists()) {
      outputFile.delete()
    }
    val out = FileOutputStream(outputFile)
    bitmapWhiteBG.compress(Bitmap.CompressFormat.JPEG, quality, out)
    bitmapWhiteBG.recycle()
    out.flush()
    out.close()

    val map = WritableNativeMap()
    map.putString("uri", Uri.fromFile(outputFile).toString())
    map.putInt("width", width)
    map.putInt("height", height)
    return map
  }

  private fun getOutputFilePrefix(filePath: String, page: Int): String {
    val tokens = filePath.split("/")
    val originalFilename = tokens[tokens.lastIndex]
    val prefix = originalFilename.replace(".", "-")
    val generator = Random()
    val random = generator.nextInt(Integer.MAX_VALUE)
    return "$prefix-thumbnail-$page-$random"
  }
}

File: ios/ReactNativePdfHelpersModule.swift
import ExpoModulesCore

public class ReactNativePdfHelpersModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ReactNativePdfHelpers')` in JavaScript.
    Name("ReactNativePdfHelpers")

    // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
    Constants([
      "PI": Double.pi
    ])

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      return "Hello world! 👋"
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("generateThumbnail") {filePath: String,page: Int, quality: Int, promise: Promise ->
      guard let fileUrl = URL(string: filePath) else {
          promise.reject("FILE_NOT_FOUND", "File \(filePath) not found", nil)
          return
      }
      guard let pdfDocument = PDFDocument(url: fileUrl) else {
          promise.reject("FILE_NOT_FOUND", "File \(filePath) not found", nil)
          return
      }
      guard let pdfPage = pdfDocument.page(at: page) else {
          promise.reject("INVALID_PAGE", "Page number \(page) is invalid, file has \(pdfDocument.pageCount) pages", nil)
          return
      }

      if let pageResult = generatePage(pdfPage: pdfPage, filePath: filePath, page: page, quality: quality) {
          promise.resolve(pageResult)
      } else {
          promise.reject("INTERNAL_ERROR", "Cannot write image data", nil)
      }
    }

    AsyncFunction("getPageCount") {filePath: String, promise: Promise ->
      guard let fileUrl = URL(string: filePath) else {
          promise.reject("FILE_NOT_FOUND", "File \(filePath) not found", nil)
          return
      }
      guard let pdfDocument = PDFDocument(url: fileUrl) else {
          promise.reject("FILE_NOT_FOUND", "File \(filePath) not found", nil)
          return
      }

      promise.resolve(pdfDocument.pageCount)
    }
  }

  func getCachesDirectory() -> URL {
      let paths = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
      return paths[0]
  }

  func getOutputFilename(filePath: String, page: Int) -> String {
      let components = filePath.components(separatedBy: "/")
      var prefix: String
      if let originalFilename = components.last {
          prefix = originalFilename.replacingOccurrences(of: ".", with: "-")
      } else {
          prefix = "pdf"
      }
      let random = Int.random(in: 0 ..< Int.max)
      return "\(prefix)-thumbnail-\(page)-\(random).jpg"
  }

  func generatePage(pdfPage: PDFPage, filePath: String, page: Int, quality: Int) -> Dictionary<String, Any>? {
      let pageRect = pdfPage.bounds(for: .mediaBox)
      let image = pdfPage.thumbnail(of: CGSize(width: pageRect.width, height: pageRect.height), for: .mediaBox)
      let outputFile = getCachesDirectory().appendingPathComponent(getOutputFilename(filePath: filePath, page: page))
      guard let data = image.jpegData(compressionQuality: CGFloat(quality) / 100) else {
          return nil
      }
      do {
          try data.write(to: outputFile)
          return [
              "uri": outputFile.absoluteString,
              "width": Int(pageRect.width),
              "height": Int(pageRect.height),
          ]
      } catch {
          return nil
      }
  }

}
